codegen/masm/src/codegen.rs

```diff
@@ -0,0 +1,273 @@
+use crate::constants::{AUX_TRACE, MAIN_TRACE};
+use crate::visitor::{
+    walk_boundary_constraints, walk_constant_bindings, walk_integrity_constraint_degrees,
+    walk_integrity_constraints, walk_periodic_columns, walk_public_inputs, AirVisitor,
+};
+use ir::{
+    constraints::{ConstraintRoot, Operation},
+    AccessType, AirIR, ConstantBinding, ConstantValueExpr, Identifier, IntegrityConstraintDegree,
+    NodeIndex, PeriodicColumn, PublicInput, TraceAccess, Value,
+};
+use std::borrow::Cow;
+use std::collections::btree_map::{BTreeMap, Entry};
+
+pub struct CodegenConfig {
+    pub ood_frame_address: u64,
+}
+
+pub struct Codegen<'ast> {
+    ins: Vec<Cow<'ast, str>>,
```

Comment:
Using `Vec<Cow<_>>` instead of `Vec<String>` because the majority of cases an allocation is not necessary. Using `Cow` allows to refer to the `&'static str` and go around memory allocations. This is neat since most of what this backend does is pushing strings to this vector.